src/db/util/conditions.ts

```diff
@@ -0,0 +1,178 @@
+/* eslint-disable @typescript-eslint/no-namespace */
+// we use namespaces here to neatly organize and scope the many type and type guard definitions
+
+import Knex = require('knex');
+
+/**
+ * Symbols to use for where condition construction
+ */
+export namespace ConditionSymbols {
+  export const BUILDER = Symbol('builder');
+  export const AND = Symbol('and');
```

Comment:
What exactly would be the benefit of using `AND`, besides completeness with having `OR`? When properties are supplied to `where`, `AND` is the default. Is there some construct where this would be useful, that I am not seeing?

If not, I would like to drop `AND` because I would not like to keep telling people during reviews: "Hey, don't use `AND` since it is a default behavior anyway, let's simplify this`

Reply:
Yeah, in cases where we want to use OR or a query builder in one of the AND clauses, or if we want to combine some query builder stuff with some normal property stuff. Something like:

`(A or B) and (B or C)` or `(A or B) and C`.

something like this in the unit tests: https://github.com/UN-OCHA/hpc_service/blob/1a0166ffb0d5fe47934de8a5cf6b51566a65f8c5/test/hpc-api-core/src/db/util/conditions.spec.ts#L129-L136

Or:

```ts
{
  [Cond.AND]: [
    {
      id: 123,
    },
    {
      [Cond.BUILDER]: (qb) => qb.whereNotNull('id'),
    },
  ],
}
```

But yeah, in the cases where we're ONLY using AND, and not using an OR or BUILDER inside of an AND clause, we'd want to use the simpler / cleaner construction.

Reply:
Ok, I understand. I didn't realize immediately that
```ts
if (OverallConditions.isQueryBuilderCondition(condition)) {
  builder.where(condition[Cond.BUILDER]);
}
```
swallows "plain" conditions, so I thought you could write the same thing as:
```ts
{
  id: 123,
  [models.v4.Cond.BUILDER]: (qb) => qb.whereNotNull('id'),
}
```